modules/transactions/src/testFixtures/java/org/apache/ignite/internal/tx/storage/state/AbstractTxStateTableStorageTest.java

```diff
@@ -0,0 +1,170 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.tx.storage.state;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertSame;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.mock;
+
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.apache.ignite.internal.configuration.storage.StorageException;
+import org.apache.ignite.internal.tx.TxMeta;
+import org.apache.ignite.internal.util.Cursor;
+import org.apache.ignite.lang.IgniteBiTuple;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Abstract class for {@link TxStateTableStorage} testing.
+ */
+public abstract class AbstractTxStateTableStorageTest {
+    private static final int PARTITION_ID = 0;
+
+    private static final int PARTITION_ID_1 = 1;
+
+    private TxStateTableStorage tableStorage;
+
+    /**
+     * Initializes the internal structures needed for tests.
+     *
+     * <p>This method *MUST* always be called in either subclass' constructor or setUp method.
+     */
+    protected final void initialize(TxStateTableStorage tableStorage) {
+        this.tableStorage = tableStorage;
+    }
+
+    @Test
+    public void testStartRebalance() throws Exception {
+        assertThrows(StorageException.class, () -> tableStorage.startRebalance(PARTITION_ID_1));
+
+        UUID txId = UUID.randomUUID();
+        TxMeta txMeta = mock(TxMeta.class);
+
+        TxStateStorage storage = tableStorage.getOrCreateTxStateStorage(PARTITION_ID);
+
+        assertThat(storage, instanceOf(TxStateStorageDecorator.class));
+
+        storage.put(txId, txMeta);
+
+        storage.lastAppliedIndex(100);
+
+        storage.flush().get(1, TimeUnit.SECONDS);
+
+        tableStorage.startRebalance(PARTITION_ID).get(1, TimeUnit.SECONDS);
+
+        TxStateStorage newStorage0 = tableStorage.getTxStateStorage(PARTITION_ID);
+
+        assertNotNull(newStorage0);
+
+        assertSame(storage, newStorage0);
+
+        assertEquals(0L, newStorage0.lastAppliedIndex());
+        assertEquals(0L, newStorage0.persistedIndex());
+
+        assertThat(getAll(newStorage0.scan()), empty());
```

Comment:
What happens with old opened cursors? I have a bad feeling that we didn't think through some corner cases. It's easy to see that these scans will work in all storages except for persistent page memory (to some extent), but there's a moment before the complete invalidation when we will have to forcefully close cursors. To a lesser degree it's also applied to regular reads.
Did you have any thoughts about the ways that storage "clients" will operate with them? Should we sit down and think some more? I'm not confident right now that we fully realize what we're doing and a potential impacts of it. Maybe backup should be fully readable and we only swap it at the end, maybe closing cursors or transparently reopening them (when it comes to RO transactions, it should be achievable).
All of that is mainly about MV storage. Here it's not that problematic, RocksDB allows to have some pretty neat workarounds, right? We should still check everything though.
Please think of everything I said and we'll discuss it later.

Reply:
Changed it